File: vito/core/src/test/java/com/facebook/fresco/vito/drawable/BitmapDrawableFactoryTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.drawable;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Color;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.util.DisplayMetrics;
import com.facebook.drawee.drawable.RoundedBitmapDrawable;
import com.facebook.fresco.vito.options.BorderOptions;
import com.facebook.fresco.vito.options.ImageOptions;
import com.facebook.fresco.vito.options.ImageOptionsDrawableFactory;
import com.facebook.fresco.vito.options.RoundingOptions;
import com.facebook.imagepipeline.image.CloseableImage;
import com.facebook.imagepipeline.image.CloseableStaticBitmap;
import java.util.HashMap;
import java.util.Map;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class BitmapDrawableFactoryTest {

  private Resources mResources;
  private DisplayMetrics mDisplayMetrics;

  private ImageOptionsDrawableFactory mDrawableFactory;

  private static final Map<String, Object> mImageExtrasRounded =
      new HashMap<String, Object>() {
        {
          put("is_rounded", true);
        }
      };
  private static final Map<String, Object> mImageExtrasNotRounded =
      new HashMap<String, Object>() {
        {
          put("is_rounded", true);
        }
      };

  @Before
  public void setup() {
    mResources = mock(Resources.class);
    mDisplayMetrics = new DisplayMetrics();
    when(mResources.getDisplayMetrics()).thenReturn(mDisplayMetrics);

    mDrawableFactory = new BitmapDrawableFactory(mResources);
  }

  @Test
  public void testCreateDrawable_whenImageUnknown_thenReturnNull() {
    final CloseableImage closeableImage = mock(CloseableImage.class);
    final ImageOptions options = mock(ImageOptions.class);

    Drawable drawable = mDrawableFactory.createDrawable(closeableImage, options);
    assertThat(drawable).isNull();
  }

  @Test
  public void testCreateDrawable_whenImageIsStaticBitmap_thenReturnBitmapDrawable() {
    final CloseableStaticBitmap closeableImage = mock(CloseableStaticBitmap.class);
    final Bitmap bitmap = mock(Bitmap.class);
    when(closeableImage.getUnderlyingBitmap()).thenReturn(bitmap);

    final ImageOptions options = mock(ImageOptions.class);

    Drawable drawable = mDrawableFactory.createDrawable(closeableImage, options);
    assertThat(drawable).isNotNull();
    assertThat(drawable).isInstanceOf(BitmapDrawable.class);
  }

  @Test
  public void
      testCreateDrawable_whenRoundAsCircleAndJavaRounding_thenReturnRoundedBitmapDrawable() {
    final CloseableStaticBitmap closeableImage = mock(CloseableStaticBitmap.class);
    final Bitmap bitmap = mock(Bitmap.class);
    when(closeableImage.getUnderlyingBitmap()).thenReturn(bitmap);

    final ImageOptions options = mock(ImageOptions.class);
    when(options.getRoundingOptions()).thenReturn(RoundingOptions.asCircle());

    ImageOptionsDrawableFactory factory = new BitmapDrawableFactory(mResources);

    Drawable drawable = factory.createDrawable(closeableImage, options);
    assertThat(drawable).isNotNull();
    assertThat(drawable).isInstanceOf(RoundedBitmapDrawable.class);
  }

  @Test
  public void testCreateDrawable_whenRoundAsCircleAndNativeRounding_thenReturnBitmapDrawable() {
    final CloseableStaticBitmap closeableImage = mock(CloseableStaticBitmap.class);
    final Bitmap bitmap = mock(Bitmap.class);
    when(closeableImage.getUnderlyingBitmap()).thenReturn(bitmap);
    when(closeableImage.getExtras()).thenReturn(mImageExtrasRounded);

    final ImageOptions options = mock(ImageOptions.class);
    when(options.getRoundingOptions()).thenReturn(RoundingOptions.asCircle());

    Drawable drawable = mDrawableFactory.createDrawable(closeableImage, options);
    assertThat(drawable).isNotNull();
    assertThat(drawable).isInstanceOf(BitmapDrawable.class);
  }

  @Test
  public void testCreateDrawable_whenRoundWithCornerRadius_thenReturnBitmapDrawable() {
    final CloseableStaticBitmap closeableImage = mock(CloseableStaticBitmap.class);
    final Bitmap bitmap = mock(Bitmap.class);
    when(closeableImage.getUnderlyingBitmap()).thenReturn(bitmap);

    final ImageOptions options = mock(ImageOptions.class);
    when(options.getRoundingOptions()).thenReturn(RoundingOptions.forCornerRadiusPx(123));
    when(closeableImage.getExtras()).thenReturn(mImageExtrasNotRounded);

    Drawable drawable = mDrawableFactory.createDrawable(closeableImage, options);
    assertThat(drawable).isNotNull();
    assertThat(drawable).isInstanceOf(RoundedBitmapDrawable.class);
    assertThat(((RoundedBitmapDrawable) drawable).getRadii())
        .isEqualTo(new float[] {123, 123, 123, 123, 123, 123, 123, 123});
  }

  @Test
  public void testCreateDrawable_whenRoundWithCornerRadii_thenReturnBitmapDrawable() {
    final CloseableStaticBitmap closeableImage = mock(CloseableStaticBitmap.class);
    final Bitmap bitmap = mock(Bitmap.class);
    when(closeableImage.getUnderlyingBitmap()).thenReturn(bitmap);

    final ImageOptions options = mock(ImageOptions.class);
    when(options.getRoundingOptions()).thenReturn(RoundingOptions.forCornerRadii(1, 2, 3, 4));
    when(closeableImage.getExtras()).thenReturn(mImageExtrasNotRounded);

    Drawable drawable = mDrawableFactory.createDrawable(closeableImage, options);
    assertThat(drawable).isNotNull();
    assertThat(drawable).isInstanceOf(RoundedBitmapDrawable.class);
    assertThat(((RoundedBitmapDrawable) drawable).getRadii())
        .isEqualTo(new float[] {1, 1, 2, 2, 3, 3, 4, 4});
  }

  @Test
  public void testCreateDrawable_whenAlreadyRoundedWithBorder_thenReturnBitmapDrawable() {
    final CloseableStaticBitmap closeableImage = mock(CloseableStaticBitmap.class);
    final Bitmap bitmap = mock(Bitmap.class);
    when(closeableImage.getUnderlyingBitmap()).thenReturn(bitmap);
    when(closeableImage.getExtras()).thenReturn(mImageExtrasRounded);

    final ImageOptions options = mock(ImageOptions.class);
    BorderOptions borderOptions = BorderOptions.create(Color.YELLOW, 10);

    when(options.getBorderOptions()).thenReturn(borderOptions);
    when(options.getRoundingOptions()).thenReturn(RoundingOptions.asCircle(false, false));
    Drawable drawable = mDrawableFactory.createDrawable(closeableImage, options);

    assertThat(drawable).isNotNull();
    assertThat(drawable).isInstanceOf(CircularBorderBitmapDrawable.class);
    assertThat(((CircularBorderBitmapDrawable) drawable).getBorder()).isEqualTo(borderOptions);
  }
}

File: vito/core/src/test/java/com/facebook/fresco/vito/drawable/BitmapDrawableFactoryTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.fresco.vito.drawable

import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.Color
import android.graphics.drawable.BitmapDrawable
import android.util.DisplayMetrics
import com.facebook.drawee.drawable.RoundedBitmapDrawable
import com.facebook.fresco.vito.options.BorderOptions
import com.facebook.fresco.vito.options.ImageOptions
import com.facebook.fresco.vito.options.ImageOptionsDrawableFactory
import com.facebook.fresco.vito.options.RoundingOptions
import com.facebook.imagepipeline.image.CloseableImage
import com.facebook.imagepipeline.image.CloseableStaticBitmap
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class BitmapDrawableFactoryTest {

  private val imageExtrasRounded: Map<String, Any> = mapOf("is_rounded" to true)
  private val imageExtrasNotRounded: Map<String, Any> = mapOf("is_rounded" to false)

  private lateinit var resources: Resources
  private lateinit var displayMetrics: DisplayMetrics
  private lateinit var drawableFactory: ImageOptionsDrawableFactory

  @Before
  fun setup() {
    resources = mock()
    displayMetrics = DisplayMetrics()
    whenever(resources.displayMetrics).thenReturn(displayMetrics)
    drawableFactory = BitmapDrawableFactory()
  }

  @Test
  fun testCreateDrawable_whenImageUnknown_thenReturnNull() {
    val closeableImage = mock<CloseableImage>()
    val options = mock<ImageOptions>()
    val drawable = drawableFactory.createDrawable(resources, closeableImage, options)
    assertThat(drawable).isNull()
  }

  @Test
  fun testCreateDrawable_whenImageIsStaticBitmap_thenReturnBitmapDrawable() {
    val closeableImage = mock<CloseableStaticBitmap>()
    val bitmap = mock<Bitmap>()
    whenever(closeableImage.underlyingBitmap).thenReturn(bitmap)
    val options = mock<ImageOptions>()
    val drawable = drawableFactory.createDrawable(resources, closeableImage, options)
    assertThat(drawable).isNotNull
    assertThat(drawable).isInstanceOf(BitmapDrawable::class.java)
  }

  @Test
  fun testCreateDrawable_whenRoundAsCircleAndJavaRounding_thenReturnRoundedBitmapDrawable() {
    val closeableImage = mock<CloseableStaticBitmap>()
    val bitmap = mock<Bitmap>()
    whenever(closeableImage.underlyingBitmap).thenReturn(bitmap)
    val options = mock<ImageOptions>()
    whenever(options.roundingOptions).thenReturn(RoundingOptions.asCircle())
    val factory = BitmapDrawableFactory()
    val drawable = factory.createDrawable(resources, closeableImage, options)
    assertThat(drawable).isNotNull
    assertThat(drawable).isInstanceOf(RoundedBitmapDrawable::class.java)
  }

  @Test
  fun testCreateDrawable_whenRoundAsCircleAndNativeRounding_thenReturnBitmapDrawable() {
    val closeableImage = mock<CloseableStaticBitmap>()
    val bitmap = mock<Bitmap>()
    whenever(closeableImage.underlyingBitmap).thenReturn(bitmap)
    whenever(closeableImage.extras).thenReturn(imageExtrasRounded)
    val options = mock<ImageOptions>()
    whenever(options.roundingOptions).thenReturn(RoundingOptions.asCircle())
    val drawable = drawableFactory.createDrawable(resources, closeableImage, options)
    assertThat(drawable).isNotNull
    assertThat(drawable).isInstanceOf(BitmapDrawable::class.java)
  }

  @Test
  fun testCreateDrawable_whenRoundWithCornerRadius_thenReturnBitmapDrawable() {
    val closeableImage = mock<CloseableStaticBitmap>()
    val bitmap = mock<Bitmap>()
    whenever(closeableImage.underlyingBitmap).thenReturn(bitmap)
    val options = mock<ImageOptions>()
    whenever(options.roundingOptions).thenReturn(RoundingOptions.forCornerRadiusPx(123f))
    whenever(closeableImage.extras).thenReturn(imageExtrasNotRounded)
    val drawable = drawableFactory.createDrawable(resources, closeableImage, options)
    assertThat(drawable).isNotNull
    assertThat(drawable).isInstanceOf(RoundedBitmapDrawable::class.java)
    assertThat((drawable as RoundedBitmapDrawable?)?.radii)
        .isEqualTo(floatArrayOf(123f, 123f, 123f, 123f, 123f, 123f, 123f, 123f))
  }

  @Test
  fun testCreateDrawable_whenRoundWithCornerRadii_thenReturnBitmapDrawable() {
    val closeableImage = mock<CloseableStaticBitmap>()
    val bitmap = mock<Bitmap>()
    whenever(closeableImage.underlyingBitmap).thenReturn(bitmap)
    val options = mock<ImageOptions>()
    whenever(options.roundingOptions).thenReturn(RoundingOptions.forCornerRadii(1f, 2f, 3f, 4f))
    whenever(closeableImage.extras).thenReturn(imageExtrasNotRounded)
    val drawable = drawableFactory.createDrawable(resources, closeableImage, options)
    assertThat(drawable).isNotNull
    assertThat(drawable).isInstanceOf(RoundedBitmapDrawable::class.java)
    assertThat((drawable as RoundedBitmapDrawable?)?.radii)
        .isEqualTo(floatArrayOf(1f, 1f, 2f, 2f, 3f, 3f, 4f, 4f))
  }

  @Test
  fun testCreateDrawable_whenAlreadyRoundedWithBorder_thenReturnBitmapDrawable() {
    val closeableImage = mock<CloseableStaticBitmap>()
    val bitmap = mock<Bitmap>()
    whenever(closeableImage.underlyingBitmap).thenReturn(bitmap)
    whenever(closeableImage.extras).thenReturn(imageExtrasRounded)
    val options = mock<ImageOptions>()
    val borderOptions = BorderOptions.create(Color.YELLOW, 10f)
    whenever(options.borderOptions).thenReturn(borderOptions)
    whenever(options.roundingOptions)
        .thenReturn(RoundingOptions.asCircle(antiAliasing = false, forceRoundAtDecode = false))
    val drawable = drawableFactory.createDrawable(resources, closeableImage, options)
    assertThat(drawable).isNotNull
    assertThat(drawable).isInstanceOf(CircularBorderBitmapDrawable::class.java)
    assertThat((drawable as CircularBorderBitmapDrawable?)?.border).isEqualTo(borderOptions)
  }
}
